### ISSUE TITLE
Numeric literals as keys in quote-props

### ISSUE THREAD

**Issue by jrvidal** [eslint/eslint#2914]
Example:
```js
// (1a) with "as-needed"
var x = ({100: 1});

// (1b) with "as-needed"
var x = ({1e2: 1});

// (1c) with "as-needed"
var x = ({'100': 1});

// (1d) with "as-needed"
var x = ({'1e2': 1});
```

Output (`ecmaFeatures` does not affect it):

```bash
  9:11  error  Unnecessarily quoted property `100` found  quote-props
```

This error corresponds to (1c). I think users should be able to configure how numeric literals are supposed to be handled with more granularity.

I propose adding a `"numbers"` option (to join the `keywords` flag introduced in #2913):

* `{numbers: "always"}`: All numeric literals should be quoted. (1c) and (1d) pass, (1a) and (1b) fail.
* `{numbers: "ambigous"}`: Numeric literals should be quoted only if the literal is not equal to its string coertion. (1a) and (1d) pass, (1b) and (1c) fail. Maybe this should be the default, in line with what we have now.
* `{numbers: "never"}`: Numeric literals should never be quoted. (1a) and (1b) pass, (1c) and (1d) fail.

**Comment by ilyavolodin** [comment]
Can you also provide the list of ecmaFeatures that are on, if any.

**Comment by jrvidal** [comment]
No `ecmaFeatures` flag affects the output.

**Comment by michaelficarra** [comment]
"ambiguous" is not like what we have now. "never" will prevent someone from ever using a key like "1e2" or "1.0". :-1:, I don't see any reason to treat numeric keys differently than any others.

**Comment by jrvidal** [comment]
Yeah, I myself don't see much usefulness in the `"never"` setting, I included it for completeness.

However, I do see the use case for the others. In particular, what we have now is slightly inconsistent. Notice that both (1b) and (1d) are accepted. Shouldn't at least one of them be reported when using "as-needed"?

**Comment by michaelficarra** [comment]
No. `1e2` does not need quotes and does not have them, and `"1e2"` *does* need quotes and *does* have them. The rule is fine as-is.

**Comment by jrvidal** [comment]
> 1e2 does not need quotes and does not have them, and "1e2" does need quotes and does have them.

Sorry, could you elaborate on this point? Doesn't this constitute the potential footgun that the very `quote-props` docs warn against?

**Comment by michaelficarra** [comment]
A key of `1e2` will create a `"100"` property. A key of `"1e2"` will create a `"1e2"` property. I'm not sure which section of the docs you're talking about. The docs mention this particular example as justification for the "always" option, but we are talking about the "as-needed" option, which is working just fine.

Remember, the "as-needed" option is essentially a sane version of your "never" suggestion, since it allows quotes only when they absolutely must be used.

**Comment by jrvidal** [comment]
> the docs mention this particular example as justification for the "always" option

I now see your point. My intention with this proposal is to allow users to define more precisely what `"as-needed"` means. I would suggest to replace my original `"never"` option with another that behaves as the default we have now. Not sure about what the appropriate name would be though.

**Comment by nzakas** [comment]
I feel like this is a stretch. @jrvidal it seems like you don't have a use case for this yourself, but you feel like someone else might? If that's the case, I'd suggest we close this issue. It's very hard to know what other people actually want or expect without them telling us. :) We can always add more options if we're wrong, but it's very hard to remove options if we're wrong.

**Comment by jrvidal** [comment]
Well, I would certainly use the `numbers: "always"` setting. My personal preference leans towards `"as-needed"`, but I wouldn't like to allow the number literals gotcha (which btw I discovered by reading the docs here).

I just feel that there are a few separate concerns mixed in `quote-props`: style, keywords and/or numbers, and it would be nice to be able to fine tune them.

Anyway, if you still think it's a stretch, feel free to close it. I have another one in the pipeline regarding getters/setters/methods. :grimacing:

**Comment by nzakas** [comment]
What is your use case? You just want all numbers to be quoted?

**Comment by jrvidal** [comment]
Yes, I would like to have the ability to specify `"as-needed"`, but enforcing quotes for number literals.

**Comment by nzakas** [comment]
Okay, how about we just keep it simple and add numbers:true as an option (false by default)?

**Comment by michaelficarra** [comment]
So `numbers: false` means "quote numbers only as needed" with the as-needed option and "quote all numbers" with the "always" option? That is really confusing. :-1: from me still.

**Comment by nzakas** [comment]
numbers:true would be to always quote numbers

**Comment by jrvidal** [comment]
> Okay, how about we just keep it simple and add numbers:true as an option (false by default)?

That works for me. I partially agree with michaelficarra that using boolean flags is not very expressive.

Maybe we could instead do:
* `numbers: "always"` (the only possibility when the main config value is `"always"`), and
* `numbers: "as-needed"` (the default when the main config value is `"as-needed"`).

If deemed a better nomenclature, I would then update #2913 for the sake of consistency (which is anyway blocked).

**Comment by nzakas** [comment]
I'd like to start with Boolean flags. Number properties never need quotes, so as-needed doesn't make sense just for numbers. It's either yes you want them quoted or no you don't, so I think booleans are the way to go.

**Comment by michaelficarra** [comment]
Except false for this option doesn't mean you don't want numeric keys quoted. It means they should only be quoted if all keys have to be quoted due to the other configuration options. That interface doesn't make sense to me.

**Comment by nzakas** [comment]
numbers is an override switch. For "always", numbers is true by default; for as-needed, numbers is false by default. In either case, you can override.

**Comment by michaelficarra** [comment]
Okay, now I understand. So the default state is magical. What confused me is that this option has 3 possible states, but you can only explicitly specify two of them. I don't like it, but at least I understand it now.

**Comment by gyandeeps** [comment]
@eslint/eslint-team I think we have discussed it but now its decision time (I guess)

**Comment by jrvidal** [comment]
I've been AFK for some time, but I can work in this now.

**Comment by nzakas** [comment]
Yeah, let's start with `numbers` as a boolean option.
